### PR TITLE
Implement letrec type derivation

### LINKED ISSUE #16: implement [Letrec] rule
implement a derivation rule [Letrec]

### PR DESCRIPTION
**Pull request by youxkei**
This PR resolves #16.

### CODE REVIEW

File: lib/derivation.ml [hunk 1]
```diff
@@ -1,42 +1,178 @@
...
+  | Abs (vs, e) ->
+     let new_tyvars = List.map ~f:(fun v -> (v, (new_tyvar ()))) vs in
+     let added_context =
+       List.fold_left
+         ~f:(fun context (v, tyvar) -> Context.add v tyvar context)
+         ~init:context
+         new_tyvars in
+     derive added_context e >>= fun (ty_e, c) ->
+     let tyvar = new_tyvar () in
+     Ok (tyvar, Eq (tyvar, TyConstraint (TyFun (List.map ~f:snd new_tyvars, ty_e), c)))
```
Comment (yoshihiro503) [inline lib/derivation.ml]: 37-45行目:
`Abs` のパターンで複数引数に対応してもらってますね。ありがとうございます

File: lib/derivation.ml [hunk 2]
```diff
@@ -1,42 +1,178 @@
...
+  | Let (v, e1, e2) ->
+     derive context e1 >>= fun (ty_e1, c1) ->
+     derive (Context.add v ty_e1 context) e2 >>= fun (ty_e2, c2) ->
+     Ok (ty_e2, Conj [c1; c2])
```
Comment (yoshihiro503) [inline lib/derivation.ml]: ![derivation rule LET](data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAaYAAABGCAYAAACHQjA1AAAAhnpUWHRSYXcgcHJvZmlsZSB0eXBlIGV4aWYAAHjaVY5RCsQwCET/PcUewagZzXFKaaE32OOvkiyh70OHQR7S9X1u+hSNhax7YACc2LAhR4bgiTI34VY752RtbZlk16QyA0Y42z601f/pisDt5o6OE6ekXS4VjZopKivXG2NLDqyk797wltMPSnksVUR/jaEAAAoEaVRYdFhNTDpjb20uYWRvYmUueG1wAAAAAAA8P3hwYWNrZXQgYmVnaW49Iu+7vyIgaWQ9Ilc1TTBNcENlaGlIenJlU3pOVGN6a2M5ZCI/Pgo8eDp4bXBtZXRhIHhtbG5zOng9ImFkb2JlOm5zOm1ldGEvIiB4OnhtcHRrPSJYTVAgQ29yZSA0LjQuMC1FeGl2MiI+CiA8cmRmOlJERiB4bWxuczpyZGY9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkvMDIvMjItcmRmLXN5bnRheC1ucyMiPgogIDxyZGY6RGVzY3JpcHRpb24gcmRmOmFib3V0PSIiCiAgICB4bWxuczpleGlmPSJodHRwOi8vbnMuYWRvYmUuY29tL2V4aWYvMS4wLyIKICAgIHhtbG5zOnRpZmY9Imh0dHA6Ly9ucy5hZG9iZS5jb20vdGlmZi8xLjAvIgogICBleGlmOlBpeGVsWERpbWVuc2lvbj0iNDIyIgogICBleGlmOlBpeGVsWURpbWVuc2lvbj0iNzAiCiAgIHRpZmY6SW1hZ2VXaWR0aD0iNDIyIgogICB0aWZmOkltYWdlSGVpZ2h0PSI3MCIKICAgdGlmZjpPcmllbnRhdGlvbj0iMSIvPgogPC9yZGY6UkRGPgo8L3g6eG1wbWV0YT4KICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIAogICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgCiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAKICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIAogICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgCiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAKICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIAogICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgCiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAKICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIAogICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgCiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAKICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIAogICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgCiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAKICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIAogICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgCiAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAKICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgIAogICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgCiAgICAgICAgICAgICAgICAgICAgICAgICAgIAo8P3hwYWNrZXQgZW5kPSJ3Ij8+Ajoo9AAAAARzQklUCAgICHwIZIgAABPjSURBVHja7Z09cOtIcsf/crmKKCdE1QVEZMGRcIFLyARHwtUFwkaCI8GRUE6IjLiI2IwZkQkXCZv4wZGwkeBIuLpAcERsRGz5qoQXCRsJe4mwkfCicfD0nqgvkuKXqKf+VSEQJZI9Pd3TPT2jmQ3GGANBEARBrAn/QCogCIIgKDARBEEQBAUmgiAIggITQRAEQVBgIgiCICgwEQRBEAQFJoIgCIICE0EQBEFQYCIIgiAoMBEEQRAEBSaCIAiCoMBEEARBUGAiCIIgCApMBEEQBAUmgiAIgqDARBAEQVBgIgjiPjWKNEYY+IiL96mBMg0RBBHSoiZzIL6dwFRmGQrqH+Kt2U4ZwZIlGEEJUTOhie+zDwRFgyLWiCwZku4jo/hEvICNtbxavfChyh6UJIcrT/rjDK4eQ4scyNSf82f6cQA/jJHXHHiOAwBwggqrZ0KuIli9Cj3fhLCuTXiR7SyaCpEuwqg9lLEJngwKqGOYoo7cyZHa00Rp8ud378MA/nEN810EloP//Q0QyukGg6IsQAnZnOach7BNB6low3NDuCJ397sigmMayLMIhZ6usUG/xHamHQAL+LoHKfKgTmWLNXhZpKD0BU6CLAJxXgIQyZ/Jh6di7Up5ZeQiExU08RvKoiRrW4XOYwuKYqM0Y6ShDXXEoAGAE3W4JofsIw9VW9889mW2M+0AWKEoS1RkJnNn8gT58NsMTFUML9XgWDIEgALTKlSe2tD0H1CbEUJLAvdc4qvqUFsqNGVdG0K2Q5APv2kfXs/AVCFxYyi2BkGUIAIoivJV8qwqC+BYFhzXg+s6cLxk/oy5CKCJIhQvmyK5zBCYMgRuAxsbTzyCiXgRiqkTOMaf8bPQQeAqzxr0Z6vmwKs6VG4tXXNtbGf8AOLDcV04lgHDiVHWOULXgeu6sO1gwgaBGkU+Q4mrKlC8xHhXZXvr7Mvr6KvfvA/fZ23WmOrURyQ78AQAECE0gU9lgQpYYT20QurqMEIJfuxDu/3i2NIRliqseQSpK1RVhaqsUQNjDKhAYNpIlB5CW4SAFI5TwfH0u3ULXoS0AOMqfAfBL8DOkQ1l4udp8AJgHW16PWzn3gjw+KXchxNJcF0VPHK4sgo902H5PqzCgqL5gGHCezabTeGaEczUw8sS3gyuk8P1tSnWvZZje69jM3P48hr66rfuw+sZmOoMfijA9r5YiwBJBFDkKKYZXIoQhpSMVTgn9xCHxtjPqiIL+vclzGHy1ZBRJYhLEca8q9mSjbSyJ6siDlFaEQL19guzELWkQ5GkRZs0ouAnfMIOTEOcboBZy6g0p+0sVJYCWQGIpvhokIy8HLpnjQSHX5ELOgwRQClDtVRYE0v/M6TevA5bMtCLFXga/wq2J0AQOVR5gXLa/liAP8/ly2vnq9+4D69rYMqDAJzljezZESAKDSAvUNZTpFyigTD1oczZyUHvR/za2kUVe/DiGjUqlKUI3ffuZSNVmSHxHLh1D6m32IItpzlwRnUTJ+AU52nnm0eOOkOaA9hUIM8wei9TByu1nUXl53kM37GRKh4iU3g0Z5AdB1/Xo+sMad6A6tyWXhQL/hJVKNkuFMuEVbnwjDFrEEuxPQ66G8A0HBg2D8/RIAvckv15el9eha/WeQQvSFEBqIoMpWjB6+kQ55VjDh9emkzfTGAqfLhxBWQWspHMsCgAfCpRTLvLdO6RJUWcA5u2B9+Rx/xZgCDnwRU5Sn7ZxfYCcVRB8vnlySFIk7PYOoUfcLBuU/rV6mCJtlNPOQeZ8Edl7KHnB4hLFW5gPFG64SCKI1pOE6SQ0VNfMg0v8fH//gvf/fM5Wv+0MUsmgY+BDggZ/KkWGRZoe6IGp5fDsmzYlQHLcWFIr+/Lq/HVFD0ngR5+CYgFfFWGakfIfXUxedOLfXgFMr3twFQg8CpYYfAoi8ndDL//foWBqa5QfwIkafyX8YoJWwGy3MHS9xCXCaJcgiUtQQ5OhiIBP1bFxLcXYYBK8V5HB0uzHQEiV6KsgLGLL3WJEsJYxxc0G75mo0osKLKKLE3hys+7dxYn+FXUoQijwbS6H7yekHfrX/8T/RlmEnUewvFK+D0bqrB62ysDHYrDwUsLGGvkyyvx1bpEkSVIckCRAUCEqomwwxQFVMwVn2f14WXKtCBedVdeGXoo9acX7QRBAFCiKFeUkQsSpFbjmV6NEaWrnxlUSYRMkCAvJYURYTh7aH6MEefj/M2FW1uw5fUqTs9vOxJ0vYTnZ2MnRHngIlONqU4h4FUbpvQzwiB7lMF7toOo+BxQk/gjGrJyNwCUIbxoSRG+TtFzK1j+C4LSQm2vRBwkqDVrNUHplXz5WX1xOsIiw93ErUaeFeAkeQH59ow+vFSZ3nhgqnMPZijBfKaswIsCmvi0wv9HUWHbErL4/nbSKgtguwXEeT0096DyPGQnnVZDSKMEtaTMbiy5B4XnINnJk4OvYIQIO4BnmAjz+mEHIfYdeLkK15LXaifPomxHsgNYqQnV9BDn9wNDlcfwLRVGpE/ehjtSsuO4BuoHqqyiHpwgQVEDdRoi47dGJmkFQreEbk7KU3nwgvjyfqh5mK71wix4AbY38llVDXArXXWf05eX6Kt16qCXqvA97a4vJ/jp2Bi8AB9+UqZ3V8orI9iWiyjJUEKC5ckI7VHHzxHaPQRxgrrRQO6bMHIDtmc9s2g5o8M+NVA5MQLHge3kkHigqioIioGeLy/giBkOHP/5mTbTLEoemqXM0TYOPM+jSCJkUJ8oAfHQvAx56sPpGYg4/u5sLV6GbvXuHWvy+tOkBdsOJ8GMUqhJAM/R4IychM0JCkw7QKrNYlv3Bwhes2EbMerIQY/X4MUaQqsHyykg84BoO5i83CTDjWZYL+GlGUozi7C9WVmMP8/ny0vy1SKE1avgJBEM4SV+Ol5fc/nwszK9MoyYiWF3i222z9+EHIP+ERt8wzpYHy7Z0U6Dtd6JTqbr//elk2e5OmWdwz47v/7848VgwK5X5KfzyPRa0H1M3zwFMghrUzsmiHdHGcGyEqi2AbEqkGcR/LB4MLtasZ9OJdN7KuW9ceo8hBekSOMSVe3BdlIohg1jxZsDppWjziJU8mKPuF8XHawfHDgOj9aY3rcP1KjrTyteY1onZSWwlH/HD78AP/zPn7++vNkZwluyn84r02uynvcxEQsiR+DmUB2dZkyr0rinQHZFhFkIXSB91JkDRYmgxNP+/xT5KUGBiSAWTIUsdOEFGSpBhtlzob/D0aZMfHhhgqzkoTs9WApFaYICE0EQBPFGoc0PBEEQxFqxkM0PSZLg73//O2mTIAjiHfHHP/4Rv/vd79YzMP31r3/F3/72N+olgiCId8T29vZSAhOtMREEQRBrBa0xTaDMMhSkBoIgiJVBM6ZxFD5U2YOS5HAnHlOWwdVjaJEDmTRHEHNQo4gD+GGMvObuzn4TVFg9E3IVwepV6PkmBJL524TON3v2ICn2Ya/JgCY7OJvm789Ze6e9gLOuBqyzCQbcPa32GbuZ+3Ov2cVgwK5uqGen54Id726yzf0P7JKUsRJuLk5Ye2eTbR8csfPL+8Z6c3nKOgcHbG+rwba6wwV94Tlrbx+w0+s3JPM7gEp5z5XwIheZqKCJ31Z49QYAKOjFAwwGAwwGZ+juAFVRYe5TbooA5r9pcBLq2xfBc+/3OJ1V+1xsQVFslGaMNLShPjgVmxN1uCaH7CMPVVtMXaIMXCRVBjfI34zM7wE6K+8pqhheqsGxOMQ//OU2MK1uAs5Lyu3R9xUKvrHgggMxPRKsKIdFili+y6U2NP0H1NYAoSU9e5gop+pQWzU0ZRHfmsELODixi9jwkFg+XnJq0uvI/D6gGdNjc0PixlBsDYIoQQRQFOX6Duh1jtDWoIg8OF6ErNkIR+4VKnwV3MYGNv7lT/gJv+HH7zawsXH78DrCahoH9OG4LhzLgOHEKOscoevAdV3YdoDsFZVTZQEcy4LjenBdB46XzH3bexUZ4Dfu9MRJDu7fSVsiNCQIggBRMuCFPixVgsBx4EUVdly+ehtQBNBEEYqXTWFDGQJThsCN2MboI5iIl9nHdQLH+DN+FjqTL2XkOPCqjkUcu1fHHmLZhiHpcLTsZbcIv5LM7waqZj6oFw/6rHNy9XXd6LAJht1jdrWyNab760Inew3W2Dt5+p6U63PW3mqwxuYe6344ZWenx6yz22Jo7rLjL4si15dsOBiw85M220KD7fbP2WAw+PwMryavXV0cs3b3/Pb7L1h/u8V29trswyVj1+dtttXYZJ3Ba/TUNRv0d9nmdpudjXTOWXufHV/NawRXbDgYsMHgnJ20txiah+zhMuPV4JSd9PdYC2BAk20fHrGTs1N2tN9iaB2ys5tXbsPFEdtpNtlWdzChjy/Zh4Nddnh0ys6HQ3YxPGb7e302uLhgF1+eJS9MXh7tsAbAdo6mW8m7WYg4l+x4d4f1L7506DHb2+mzi7WW+f1ApbwHmaMfCrC9L2U7AZIIoMhRTFPMK0IYUjI2e+LkHuLQWEhhMO2Z+KFU8SGLYYqfX9M0EZX4HVw3heUrAC9CVkRASMGDgyCrUKYuKVSIvBy6Z43c+vkrckGHIQIoZaiWCmts6bxAYOhwXzKt4lV4sQ+NHzersaB/X8IcJtC+KLNKEJcijJH3VWWGxHPg1j2k3pQN5wTIt4eOCunTQgiKDoNL4Hz/F7QOQiSB9llHXAz3DwmSHJi0pDBNG+o8ghekqABURYZStOD1dEy8WFiykVb2FLOGEKUVIfhyhW4WopZ0KJL0YGI+oxyTp3aIgp/wCTswDXG67uHm7F8AyHwEvI34SzMFE7akwEvsKU5Bn13m5enx24IC0wh5EICzvJGj5wWIQgPIC5Q1MPEWLdFAmPpYTSk5RRj9goZiQ6wyZF8rNhwkqYH/zlIUUOY8Rp+D7Dh3TlNnSPMGVOe2dKFY8Cc2VoQZZjAX2vYCQe9H/NraRRV78OIaNSqUpQjd975eo16lAYKcB1fkKPll1aJa0CztLnDzHDjUU9zJNE0bUvScBHr45ecCvipDtSPkvrqQS904zYEz6gNxAk5xHtna0uSoM6Q5gE0F8guztdn7t0LkxpCd3kjCxUGzNfR6IUp1wpbumWVefn9SYPrWKHy4cQVk1sh6Qo2iAPCpRFEC63VZSoWyBD798if84S+Pf9vYXciwBVEc8bw0QQoZPZV/5aaniHNg0/bgO89PS3jFhK0AWe5g/kWb53XE80tqQ12iyD7PvhT5c5BXNRF2mKKACmnxToA4qiD5/OrlEKTJVYQ6hR9wsG6n6DP3bxHCK00EygMfkW1YtY4gN+FIS5B55f1JgemtRyUEXgUrDPDQVnM3w++/X8fAJEAUgaZ8hjLUZsq2iiRELujQpOnencUJfhV13F2tU6MoqvvBaxXUFepPgCSJb9fkpmkDpyMs9NE3Ic8KcJK8HFMsE0S5BEtaoRycDEUCfqyKibGlCANUyvx3rKZeAMGOn5BdgGFLUNwYdjDGp2aVedX9+YahXXkAytBDqduPghIACIIAoERRrtu+PBmGvoXfIge9tL5n7JmnQtID3NsbxnH4fM11/bWs4Jn/AcvPx2b1nu0gKj4H7yT+iIas3GV2ZTjFTqYCgSFDkqTpH8VCXI3PVKXWM9voixhR+gY2xc/Qhjp10EtV+J62lLJPlUTIBAkyN2kSsEg5RBjOHpofY8T5uJjpwq0t2PKc31hF8FIFtv70NJfTHBi5h7BcvszL7s83Df2n+RHb2x+z6+78kDWn2n2zuF1515fD251z56y7A4adLju/3Uk3vLwe3ULI+jtNhuY2O+ifsLPzM3bS32ebjQbb7g4f7MYasu4WWHOnzT6cnbLj9g5rYpN1Bs9vF7r+sMcazR12dPF5t+Lh7hZrHZzefu4lO+n02fn16/TbRX+btQ5O7+1WvB5+YJ32MRs+aNKwu8U22+cv6YDbXXkDdtreYmjssaPzARsMhl93qN1cDdngwwFrocUOPgzY8OqGsevLx68tqA3s8oQd7h2yk2mPoLg4YrvNJtvuTmuRN+zsoMka+6fjd/GNk+PiiO00G2yrc/7Ck0qu2Vlnm7W2D9nJxc1DB2Vnx13WPR6w50ztJf17ebTLdo/HK/Hqwx7bmXhKw3wyv7g/3xnvNzBdnbLO/g7bbDZYo7nN9o8ebqu9YCedA7a31WKNRoM1t3bZQfuYPT+OD1l3v8/mP3Tk8ZFE957NzoPgd8XO+gdsd7vFmo0m29zaZYfPOMTN8AM73GmxZqPBWlt7rHN6OX4AuTpj3XaH9ftd1j0+Z9c3Q3Z8uM8Ou0fsqH/ETl/Vqa7YWfeQHXb7rN/vs263y45Oh0+2+6WB6aK/zRrP6L+5f8Ku2SU73m3ce72xe8xOO1v33tfYPZ5wlNGUbbg6ZZ3DuyTgYjBmwBvZ5r+32WI7/Wkt8pId726y/ZOr8T4zTo7b72xtd2ZK0K4Hx6x9sM8ODg9Zu91m7XabdbrHj475mb1/L9jR/iE7m6S8m3PW2etO5cszyTxLf74z6BBX4psncyTolY/CV99gnTmCZSXQejZkHqirDH5Qw/WM1ZZ/XiBH6nqAY0Oh/l3f/qRSHkG8Vpn2hPW7Hba/3WTNrX3W6fbZyfAN/afjzTlrPzF73uwM11iOS3bcP2FX1L/r2580YyII4j1RZx680oSj8aQMYmZoVx5BEAsiRxiLMCgoEXNCMyaCIAiCZkwEQRAEQYGJIAiCoMBEEARBEBSYCIIgCApMBEEQBEGBiSAIgqDARBAEQRAUmAiCIAgKTARBEATxWvw/NHBi5FCmUqMAAAAASUVORK5CYII=)

LET規則も実装くださってますね。ありがとうございます。

File: lib/derivation.ml [hunk 3]
```diff
@@ -1,42 +1,178 @@
...
+         ~init:context
+         new_tyvars in
+     let constraints_result =
+       List.fold_left
+         ~f:(fun constraints_result (v, f, tyvar) ->
+               constraints_result >>= fun constraints ->
+               derive added_context f >>= fun (ty, c) ->
+               Ok (Eq (tyvar, ty) :: c :: constraints))
+         ~init:(Ok [])
+         new_tyvars in
```
Comment (yoshihiro503) [inline lib/derivation.ml]: `let constraints_result = ...` ここ ResultのmapMがあるともうちょっと簡単に書けるんですけどね

```
let constraints_result =
  Result.mapM ~f:(fun (v, f, tyvar) -> derive added_context f) new_tyvars
  |> List.map ~f:(fun (ty', c, tyvar) ->[Eq (tyvar, ty'); c])
  |> List.concat
in
```
みたいに。

File: lib/derivation.ml (hunk 2 above)
Comment (yoshihiro503) [reply lib/derivation.ml]: Thank you for implementing the case of the rule [LET] 
[just a comment]

**Comment by youxkei** [comment]
@yoshihiro503 I moved unit tests for derivation.ml into test directory

**Pull request merged**